libs/knowledge-store/ragstack_knowledge_store/langchain/cassandra.py

```diff
@@ -0,0 +1,198 @@
+    def __init__(self, embeddings: Embeddings):
+        self.embeddings = embeddings
+
+    def embed_texts(self, texts: List[str]) -> List[List[float]]:
+        return self.embeddings.embed_documents(texts)
+
+    def embed_query(self, text: str) -> List[float]:
+        return self.embeddings.embed_query(text)
+
+    async def aembed_texts(self, texts: List[str]) -> List[List[float]]:
+        return await self.embeddings.aembed_documents(texts)
+
+    async def aembed_query(self, text: str) -> List[float]:
+        return await self.embeddings.aembed_query(text)
+
+
+def _row_to_document(row) -> Document:
+    return Document(
+        page_content=row.text_content,
+        metadata={
+            knowledge_store.CONTENT_ID: row.content_id,
+            "kind": row.kind,
+        },
+    )
+
+
+def _results_to_documents(results: Optional[ResponseFuture]) -> Iterable[Document]:
+    if results:
+        for row in results:
+            yield _row_to_document(row)
+
+
+class CassandraKnowledgeStore(KnowledgeStore):
+    def __init__(
+        self,
+        embedding: Embeddings,
+        edge_extractors: List[EdgeExtractor],
+        *,
+        node_table: str = "knowledge_nodes",
+        edge_table: str = "knowledge_edges",
+        session: Optional[Session] = None,
+        keyspace: Optional[str] = None,
+        setup_mode: SetupMode = SetupMode.SYNC,
+        concurrency: int = 20,
+    ):
+        """
+        Create the hybrid knowledge store.
+        Parameters configure the ways that edges should be added between
+        documents. Many take `Union[bool, Set[str]]`, with `False` disabling
+        inference, `True` enabling it globally between all documents, and a set
+        of metadata fields defining a scope in which to enable it. Specifically,
+        passing a set of metadata fields such as `source` only links documents
+        with the same `source` metadata value.
+        Args:
+            embedding: The embeddings to use for the document content.
+            edge_extractors: Edge extractors to use for linking knowledge chunks.
+            concurrency: Maximum number of queries to have concurrently executing.
+            setup_mode: Mode used to create the Cassandra table (SYNC,
+                ASYNC or OFF).
+        """
+        self._embedding = embedding
+        self._session = check_resolve_session(session)
+        _keyspace = check_resolve_keyspace(keyspace)
+        _setup_mode = getattr(knowledge_store.SetupMode, setup_mode.name)
+
+        self.store = knowledge_store.KnowledgeStore(
+            embedding=EmbeddingModelAdapter(embedding),
+            edge_extractors=edge_extractors,
+            node_table=node_table,
+            edge_table=edge_table,
+            session=self._session,
+            keyspace=_keyspace,
+            setup_mode=_setup_mode,
+            concurrency=concurrency,
+        )
+
+        self._query_by_embedding = self._session.prepare(
```

Comment:
I feel like the query by embedding could/should be in the `knowledge_store.py`. We could have a "get nearest nodes", and then have LangChain wrap that.

Reply:
done.